### ISSUE TITLE
Proposal: attempt to do background/unattended installs on application exits

### ISSUE THREAD

**Issue by gregneagle** [munki/munki#801]
Since there is now a Munki tool that monitors app launches and exits (app_usage_monitor) -- perhaps this tool could be extended for this scenario:

Google Chrome is an unattended install, but a user has Chrome open nearly all the time, and so Munki cannot install it unattended.

What if app_usage_monitor, when it sees that the user quit Chrome (for whatever reason), could then tell Munki to attempt its unattended_installs?

It would probably _not_ be a great idea to do this if the reason Chrome just quit is because a logout or restart or shutdown has been initiated, so we'd need to see if we can detect when we are in any of those states before triggering an install of unattended items.

**Comment by gregneagle** [comment]
This https://developer.apple.com/documentation/appkit/nsworkspacewillpoweroffnotification?language=objc looks promising for determining if the user has initiated a logout/restart/shutdown.

**Comment by erikng** [comment]
I think munki would have to be extended in order to install unattended installs separately, but this sounds like a great feature.

Thanks,
Erik Gomez
________________________________

**Comment by gregneagle** [comment]
Yes, there would have to be changes to Munki itself, but not too many, really.

**Comment by gregneagle** [comment]
More thoughts (just trying to capture these):

We have a list of pending updates in InstallInfo.plist, some of which are marked as unattended_installs with blocking_applications.

When an application quits, app_usage_monitor is notified. It can find all the pending unattended_install blocking_applications. If the app that just exited is in the list of blocking applications, it could then run munkilib.installer.run(only_unattended=True)

**Comment by keeleysam** [comment]
I like this idea, but for software other than Chrome. However, I think this could be adapted for some special software like that.  In our environment, we configure Chrome to auto update itself, which it does itself when quit.

There's already some contention between Munki and Chrome here - I've seen it where Chrome is running an old version, Munki has a new one, and Chrome itself has an even newer update.  In this case, because Munki doesn't actually check again to see if an item needs to be installed before installing it, Chrome would be updated to the newest version and then downgraded by Munki.  After that, Chrome would want to update itself to the latest version again.

For Chrome, I might actually like the ability to have Munki check for the updates but not actually do them, or something like that.

I do like this for software which doesn't do that, such as Microsoft products or Apple apps.

**Comment by gregneagle** [comment]
Chrome is just an example of software that might be open a lot. I would not implement this exclusively for Chrome; that would be crazy.

**Comment by keeleysam** [comment]
Yes, what I'm suggesting is that Munki might check for changes on the application before doing the install run, because the application itself might just have done it for you.

**Comment by gregneagle** [comment]
For your scenario, just tell Munki to not care about the version currently installed. Make the installs array like this:

```xml
	<key>installs</key>
	<array>
		<dict>
			<key>path</key>
			<string>/Applications/Google Chrome.app/Contents/MacOS/Google Chrome</string>
			<key>type</key>
			<string>file</string>
		</dict>
	</array>
```

If Chrome is not present, Munki will install the latest version it has. If Chrome is present, it will leave it alone.
No need for Munki code changes or additional complexity.

**Comment by arubdesu** [comment]
I just noticed this and it looks like it was a great idea, if I can just offer my plus-one.
